src/StreamJsonRpc/HeaderDelimitedMessageHandler.cs

```diff
@@ -203,50 +206,58 @@ private enum HeaderParseState
-                headerValueBytesLength = HeaderEncoding.GetBytes(contentTypeHeaderValue, 0, contentTypeHeaderValue.Length, this.sendingHeaderBuffer, 0);
-                sendingBufferStream.Write(this.sendingHeaderBuffer, 0, headerValueBytesLength);
-                sendingBufferStream.Write(CrlfBytes, 0, CrlfBytes.Length);
-            }
+                this.sendingBufferStream.Write(ContentLengthHeaderName, 0, ContentLengthHeaderName.Length);
+                this.sendingBufferStream.Write(HeaderKeyValueDelimiter, 0, HeaderKeyValueDelimiter.Length);
+                int headerValueBytesLength = HeaderEncoding.GetBytes(contentBytesLengthString, 0, contentBytesLengthString.Length, this.sendingHeaderBuffer, 0);
+                this.sendingBufferStream.Write(this.sendingHeaderBuffer, 0, headerValueBytesLength);
+                this.sendingBufferStream.Write(CrlfBytes, 0, CrlfBytes.Length);
+
+                // Transmit the Content-Type header, but only when using a non-default encoding.
+                // We suppress it when it is the default both for smaller messages and to avoid
+                // having to load System.Net.Http on the receiving end in order to parse it.
+                if (DefaultContentEncoding.WebName != contentEncoding.WebName || this.SubType != DefaultSubType)
+                {
+                    this.sendingBufferStream.Write(ContentTypeHeaderName, 0, ContentTypeHeaderName.Length);
+                    this.sendingBufferStream.Write(HeaderKeyValueDelimiter, 0, HeaderKeyValueDelimiter.Length);
+                    var contentTypeHeaderValue = $"application/{this.SubType}; charset={contentEncoding.WebName}";
+                    headerValueBytesLength = HeaderEncoding.GetBytes(contentTypeHeaderValue, 0, contentTypeHeaderValue.Length, this.sendingHeaderBuffer, 0);
+                    this.sendingBufferStream.Write(this.sendingHeaderBuffer, 0, headerValueBytesLength);
+                    this.sendingBufferStream.Write(CrlfBytes, 0, CrlfBytes.Length);
+                }
 
-            // Terminate the headers.
-            sendingBufferStream.Write(CrlfBytes, 0, CrlfBytes.Length);
+                // Terminate the headers.
+                this.sendingBufferStream.Write(CrlfBytes, 0, CrlfBytes.Length);
 #pragma warning restore VSTHRD103 // Call async methods when in an async method
 
-            // Either write both the header and the content, or don't write anything.
-            // If we write only the header when the cancellation comes, that would confuse the recieving side
-            // and corrupt the data data it reads.
-            cancellationToken.ThrowIfCancellationRequested();
+                // Either write both the header and the content, or don't write anything.
+                // If we write only the header when the cancellation comes, that would confuse the recieving side
+                // and corrupt the data data it reads.
+                cancellationToken.ThrowIfCancellationRequested();
```

Comment:
How likely it is that we would get a cancellation here between starting to send a message and preparing the headers? Would the savings by removing the intermediate buffere be worth not supporting cancellation here?

Reply:
The odds of cancellation are slim. But I think even if we removed this line, we would keep the temporary buffer. The reason being that writing many tiny bytes to a Stream of an unknown transport can actually be inefficient. It can be better to be chunky when writing to a Stream. When we switch to Pipes this won't be an issue, but that's a 2.0 feature.